### PR TITLE
QAT support for facial landmark detection task

### PR DESCRIPTION
**Pull request by yrajas**
# Pull Request Template

## Description

Add quantization aware training support for facial landmark detection task models.

Fixes: # (issue)

## Changes

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**Configuration**:
* Archai version: 1.0.0
* Environment:
* OS:

## Final checks:

- [ ] Follows guidelines of project
- [ ] Self-review of code
- [ ] Commented code in hard-to-understand areas
- [ ] Corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Dependent changes have already been merged
- [ ] Checked code correction any misspellings

### CODE REVIEW

File: tasks/facial_landmark_detection/README.md [hunk 1]
```diff
@@ -57,6 +57,16 @@ The trained models will be saved in the models directory. You can modify the out
 The training using the parameters in train_candidate_models.py produces another CSV file (search_results_with_full_validation_error.csv) with validation error data added from the training. The following graph is produced with such data:
 ![Results of full training](Full_Training_Validation_Error_vs_onnx_latency_(ms).png)
 
+## Quantization Aware Training
+To perform QAT on a selected model, use the --qat option which utilizes the quantizable version of the model to perform the training. 
+If needed for higher accuracy, use the --qat_skip_layers <# of layers> to partially quantize the model with specified number of layers left in floating point.
```
Comment (wchen-github) [inline tasks/facial_landmark_detection/README.md:62]: Do we need to specify the direction of skipping? I assume the skipping starts from the head of the model and go towards the stem.

File: tasks/facial_landmark_detection/README.md [hunk 2]
```diff
@@ -57,6 +57,16 @@ The trained models will be saved in the models directory. You can modify the out
...
+
+```bash
+torchrun --nproc_per_node=4 train.py --data-path $data_dir --output_dir $output_dir --nas_search_archid $arch_id --search_result_csv $csv \
+    --train-crop-size 128 --epochs 100 --search_result_archid 3484fc6b \
+    --batch-size 32 --lr 0.001 --opt adamw --lr-scheduler steplr --lr-step-size 100 --lr-gamma 0.5 -wd 0.00001 --qat
+```
+
```
Comment (wchen-github) [inline tasks/facial_landmark_detection/README.md:69]: Do you have any example results ready to show the effect of quantization on the validation error. Also curious with this task, whether the FC layer needs to be skipped or not.

File: tasks/facial_landmark_detection/README.md [hunk 3]
```diff
@@ -57,6 +57,16 @@ The trained models will be saved in the models directory. You can modify the out
 The training using the parameters in train_candidate_models.py produces another CSV file (search_results_with_full_validation_error.csv) with validation error data added from the training. The following graph is produced with such data:
 ![Results of full training](Full_Training_Validation_Error_vs_onnx_latency_(ms).png)
 
+## Quantization Aware Training
+To perform QAT on a selected model, use the --qat option which utilizes the quantizable version of the model to perform the training. 
+If needed for higher accuracy, use the --qat_skip_layers <# of layers> to partially quantize the model with specified number of layers left in floating point.
+
+```bash
+torchrun --nproc_per_node=4 train.py --data-path $data_dir --output_dir $output_dir --nas_search_archid $arch_id --search_result_csv $csv \
+    --train-crop-size 128 --epochs 100 --search_result_archid 3484fc6b \
```
Comment (wchen-github) [inline tasks/facial_landmark_detection/README.md:66]: Why is search_result_archid specified twice?

File: tasks/facial_landmark_detection/train.py [hunk 4]
```diff
@@ -148,6 +150,22 @@ def load_data(traindir, args):
...
+def setup_qat(model: nn.Module) -> None:
+    assert model is not None
+
+    _replace_relu(model)
+
+    model.eval()
+
+    # Modify quantization engine as appropriate for the target platform
+    model.setup_qconfig('fbgemm')
+    model.classifier.qconfig = None
```
Comment (wchen-github) [inline tasks/facial_landmark_detection/train.py:162]: It seems the default for classifier.qconfig is already None. Should we make this clear? See the comment in readme as well.

File: tasks/facial_landmark_detection/train.py [hunk 5]
```diff
@@ -148,6 +150,22 @@ def load_data(traindir, args):
...
+
+    model.fuse_model(is_qat=True)
```
Comment (wchen-github) [inline tasks/facial_landmark_detection/train.py:164]: Is it really necessary to have the is_qat argument? It seems like we always call _fuse_model with True.

**Comment by wchen-github** [comment]
Also, can you do me a couple of favors? Looks like I left line #97 in model.py for no reason. Then, on line 74 in model.py, maybe we should take away the devault value. For the context of this task, 1000 makes no sense and we must pass in an explicit num_classes.

**Comment by yrajas** [comment]
[like]  Yoganand Rajasekaran reacted to your message:
________________________________

File: tasks/facial_landmark_detection/README.md (hunk 1 above)
Comment (yrajas) [reply tasks/facial_landmark_detection/README.md:62]: Fixed.

**Review by lovettchris — APPROVED** [review]

File: tasks/facial_landmark_detection/quantizable_model.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
+from typing import Any, Optional
```
Comment (wchen-github) [inline tasks/facial_landmark_detection/quantizable_model.py:1]: Do we need copyright statement for this file as well?

**Review by wchen-github — APPROVED** [review]
